Add description for .spec.application.containers (#119)
k8s-service-bindings/spec: README.md

Restricting service binding to resources within the same namespace is strongly **RECOMMENDED**.  Cross-namespace service binding **SHOULD** be secured appropriately by the implementor to prevent attacks like privilege escalation and secret enumeration.

> A Service Binding resource **MUST** define a `.spec.application` which is an `ObjectReference`-like declaration to a `PodSpec`-able resource.  A `ServiceBinding` **MAY** define the application reference by-name or by-[label selector][ls]. A name and selector **MUST NOT** be defined in the same reference.  The `ServiceBinding` **MAY** define `.spec.application.containers` to describe which containers in the pod is used for binding.  If `.spec.application.containers` is not defined, all containers **MUST** be used for binding.  If the value is a positive index number, starting from 0, the container in `.spec.template.spec.containers[*]` at that index **MUST** be used for binding.  If the value is a string, the binding **MUST** be performed for the containers matching name in both `.spec.template.spec.containers[*]` and `.spec.template.spec.initContainers[*]`. A Service Binding resource **MUST** define a `.spec.service` which is an `ObjectReference`-like declaration to a Provisioned Service-able resource.  Extensions and implementations **MAY** allow additional kinds of applications and services to be referenced.

Reviewer: Moving into a separate paragraph to more clearly define the scope of what is optional and the default behavior. Using bullets to describe the behavior for each containers item. Refining the JSON Path expressions for how containers are matched.

```suggestion
A Service Binding resource **MUST** define a `.spec.application` which is an `ObjectReference`-like declaration to a `PodSpec`-able resource.  A `ServiceBinding` **MAY** define the application reference by-name or by-[label selector][ls]. A name and selector **MUST NOT** be defined in the same reference.  A Service Binding resource **MUST** define a `.spec.service` which is an `ObjectReference`-like declaration to a Provisioned Service-able resource.  Extensions and implementations **MAY** allow additional kinds of applications and services to be referenced.

The Service Binding resource **MAY** define `.spec.application.containers`, as a list of integers or strings, to limit which containers in the application are bound.  Binding to a container is opt-in, unless `.spec.application.containers` is undefined then all containers **MUST** be bound.  For each item in the containers list:
- if the value is an integer (`${containerInteger}`), the container matching by index (`.spec.template.spec.containers[${containerInteger}]`) **MUST** be bound. Init containers **MUST NOT** be bound
- if the value is a string (`${containerString}`), a container or init container matching by name (`.spec.template.spec.containers[?(@.name=='${containerString}')]` or `.spec.template.spec.initContainers[?(@.name=='${containerString}')]`) **MUST** be bound
- values that do not match a container or init container **SHOULD** be ignored
```

A Service Binding resource **MUST** define a `.spec.application` which is an `ObjectReference`-like declaration to a `PodSpec`-able resource.  A `ServiceBinding` **MAY** define the application reference by-name or by-[label selector][ls]. A name and selector **MUST NOT** be defined in the same reference.  A Service Binding resource **MUST** define a `.spec.service` which is an `ObjectReference`-like declaration to a Provisioned Service-able resource.  Extensions and implementations **MAY** allow additional kinds of applications and services to be referenced.

> The Service Binding resource **MAY** define `.spec.application.containers`, as a list of integers or strings, to limit which containers in the application are bound.  Binding to a container is opt-in, unless `.spec.application.containers` is undefined then all containers **MUST** be bound.  For each item in the containers list:

Author: ```suggestion
The Service Binding resource **MAY** define `.spec.application.containers`, as a list of integers or strings, to limit which containers in the application are bound.  Binding to a specific container or containers is opt-in. If `.spec.application.containers` is undefined, the reconciler **MUST** bind all containers.  For each item in the containers list:
```

Author: @scothis Let me know if the rephrasing makes the sentence more clear.

Reviewer: this sentence doesn't stand on it's own anymore:
> Binding to a specific container or containers is opt-in.

Since the default behavior is to bind all containers, that's not opt-in.

Author: I think that sentence can be removed because the next sentence is sufficient.

> If `.spec.application.containers` is undefined, the reconciler **MUST** bind all containers.

Reviewer: Without that sentence a conforming implementation could always bind all containers and completely ignore the content of `.spec.application.containers`.

Author: Ok, I am withdrawing my suggestion. Thanks :+1:
